### PR TITLE
Add support for ghc 9.0.2

### LINKED ISSUE #2541: Please, support GHC 9.0.2
This report is a notification & tracking report for https://www.haskell.org/ghc/blog/20211225-ghc-9.0.2-released.html support

Released on `2021-12-25`:

Complete release notes: https://downloads.haskell.org/ghc/9.0.2/docs/html/users_guide/9.0.2-notes.html

---

To have a glance for HLS: quoting

<details><summary>GHC 9.0.2 summary release notes</summary>

</p>

GHC 9.0.2 adds first-class AArch64/Darwin support using the LLVM backend, as well as fixing a number of critical correctness bugs with the 9.0.1 release, along with numerous improvements to compiler performance and memory usage:
  * [ ] First-class support for Apple M1 hardware using GHC’s LLVM ARM backend.
  * [ ] Fix a number of incorrect rewrite rules for `fromInteger` ([GHC#19345](https://gitlab.haskell.org/ghc/ghc/-/issues/19345), [GHC#20066](https://gitlab.haskell.org/ghc/ghc/-/issues/20066)).
  * [ ] Fix a number of runtime system and GC bugs which typically manifested in crashes ([GHC#19417](https://gitlab.haskell.org/ghc/ghc/-/issues/19417), [GHC#19297](https://gitlab.haskell.org/ghc/ghc/-/issues/19297), [GHC#19645](https://gitlab.haskell.org/ghc/ghc/-/issues/19645), [GHC#19667](https://gitlab.haskell.org/ghc/ghc/-/issues/19667), [GHC#18033](https://gitlab.haskell.org/ghc/ghc/-/issues/18033), [GHC#20132](https://gitlab.haskell.org/ghc/ghc/-/issues/20132), [GHC#19715](https://gitlab.haskell.org/ghc/ghc/-/issues/19715), [GHC#20399](https://gitlab.haskell.org/ghc/ghc/-/issues/20399), [GHC#20414](https://gitlab.haskell.org/ghc/ghc/-/issues/20414), [GHC#20649](https://gitlab.haskell.org/ghc/ghc/-/issues/20649), [GHC#19147](https://gitlab.haskell.org/ghc/ghc/-/issues/19147), [GHC#20051](https://gitlab.haskell.org/ghc/ghc/-/issues/20051), [GHC#20093](https://gitlab.haskell.org/ghc/ghc/-/issues/20093))
  * [ ] Fixes for various compiler performance and memory usage issues ([GHC#20134](https://gitlab.haskell.org/ghc/ghc/-/issues/20134), [GHC#13586](https://gitlab.haskell.org/ghc/ghc/-/issues/13586), [GHC#20509](https://gitlab.haskell.org/ghc/ghc/-/issues/20509), [GHC#19668](https://gitlab.haskell.org/ghc/ghc/-/issues/19668), [GHC#19996](https://gitlab.haskell.org/ghc/ghc/-/issues/19996), [GHC#19471](https://gitlab.haskell.org/ghc/ghc/-/issues/19471))
  * [ ] Improvements and bug fixes for the code generator and optimiser ([GHC#18638](https://gitlab.haskell.org/ghc/ghc/-/issues/18638), [GHC#19581](https://gitlab.haskell.org/ghc/ghc/-/issues/19581), [GHC#19700](https://gitlab.haskell.org/ghc/ghc/-/issues/19700), [GHC#19996](https://gitlab.haskell.org/ghc/ghc/-/issues/19996), [GHC#19181](https://gitlab.haskell.org/ghc/ghc/-/issues/19181), [GHC#19822](https://gitlab.haskell.org/ghc/ghc/-/issues/19822))
  * [ ] Fixes for typechecker bugs ([GHC#19410](https://gitlab.haskell.org/ghc/ghc/-/issues/19410) , [GHC#19191](https://gitlab.haskell.org/ghc/ghc/-/issues/19191), [GHC#19522](https://gitlab.haskell.org/ghc/ghc/-/issues/19522), [GHC#19400](https://gitlab.haskell.org/ghc/ghc/-/issues/19400), [GHC#19849](https://gitlab.haskell.org/ghc/ghc/-/issues/19849))
  * [ ] Bug fixes for the frontend and parser ([GHC#19397](https://gitlab.haskell.org/ghc/ghc/-/issues/19397), [GHC#19838](https://gitlab.haskell.org/ghc/ghc/-/issues/19838), [GHC#19518](https://gitlab.haskell.org/ghc/ghc/-/issues/19518), [GHC#19777](https://gitlab.haskell.org/ghc/ghc/-/issues/19777), [GHC#20242](https://gitlab.haskell.org/ghc/ghc/-/issues/20242), [GHC#19786](https://gitlab.haskell.org/ghc/ghc/-/issues/19786))
  * [ ] Emit warnings if Unicode Bidirectional Formatting characters are found in the source, possibly misleading code reviewers ([GHC#20263](https://gitlab.haskell.org/ghc/ghc/-/issues/20263), vulnerability website).
  * [ ] Many packaging-related fixes, including versioned `ghc-pkg` executables ([GHC#20087](https://gitlab.haskell.org/ghc/ghc/-/issues/20087)), and actually distributing GHC versions linked against the `native` big integer backend ([GHC#18967](https://gitlab.haskell.org/ghc/ghc/-/issues/18967), [GHC#19953](https://gitlab.haskell.org/ghc/ghc/-/issues/19953)) on both Windows and Alpine Linux. Some previous releases were still linked against the `GMP` library due to a misconfiguration of the builders.
  * [ ] A significant refactoring of `process` fixing numerous bugs mostly on Apple platforms ([GHC#19994](https://gitlab.haskell.org/ghc/ghc/-/issues/19994), process refactoring).
 * [x] Bug fixes for the linker on Darwin platforms ([GHC#20004](https://gitlab.haskell.org/ghc/ghc/-/issues/20004), [GHC#19968](https://gitlab.haskell.org/ghc/ghc/-/issues/19968), [GHC#19950](https://gitlab.haskell.org/ghc/ghc/-/issues/19950)).

</details>

---

A number of things to date (*2022-01-07*) is still not done: https://gitlab.haskell.org/ghc/ghc/-/issues/19632#note_400751, for example, boot `ghc-*` package versions not released to Hackage.

### PR DESCRIPTION
**Pull request by jneira**
* It misses the nix config
* We have to check the gitlab build
* Closes #2541 

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2567"></a>

### CODE REVIEW

File: .github/workflows/hackage.yml
Comment (michaelpj) [inline .github/workflows/hackage.yml:114]: Just while I'm here: for this and the above step, don't we already have modified `cabal.project` files for exactly this kind of thing?

File: .github/workflows/test.yml
Comment (michaelpj) [inline .github/workflows/test.yml:152]: Can we define a `cabal test` command that abstracts this stuff out. Ouch...

File: .github/workflows/test.yml
Comment (michaelpj) [inline .github/workflows/test.yml:150]: I wonder what would happen if we did this unconditionally but set the test suite to not be buildable on GHC 9 :thinking: might error, might just succeed...

File: .github/workflows/hackage.yml
Comment (jneira) [reply .github/workflows/hackage.yml:114]: well, it is a long story:
* we have to avoid any cabal.project tweaks (`allow-newer`, `source-repository-packages`) as hackage doesnt see them and without them the hackage build fails
* we want to make the packages tests and benchmarks work in hackage
* cabal solver needs all packages (including tests and benchmarks as we want enable them in general) be "solvable" although you dont build them effectively (like brittany for ghc-9.0.1)

* so we have to replicate the minimum possible of cabal.project tweaks in a controlled way to make the build work

But maybe there is another way to do it, o revise the initial premises

File: .github/workflows/test.yml
Comment (jneira) [reply .github/workflows/test.yml:152]: maybe with a shell routine? or you are thinking in other possiblesolution?

File: .github/workflows/test.yml
Comment (jneira) [reply .github/workflows/test.yml:150]: hmm i think it is good to build it even if the tests are gonna fail, or it will rot even with compile errors.
We are building all the test suites (with `tests: true`) and cabal doesnt have `runnable: fase` 😉

**Comment by fendor** [comment]
There is no release of hie-bios with support for GHC 9.0.2 yet. Going to make one now.

File: .github/workflows/hackage.yml
Comment (michaelpj) [reply .github/workflows/hackage.yml:114]: Right, but this is for GHC 9, right? Where we *do* have `cabal.project` tweaks. Basically, I don't see how adding them here can be any better than adding them in `cabal-ghc9.project` or whatever?

File: .github/workflows/test.yml
Comment (michaelpj) [reply .github/workflows/test.yml:152]: Oh hm, we would actually need a function here. I was thinking you could do something like
```
echo "CABAL_TEST=cabal test ..." >> $GITHUB_ENV
```
but I don't know if you can do something like that for bash functions :thinking:

Comment (michaelpj) [reply .github/workflows/test.yml:152]: Ugly version would be to commit a small script that does that.

File: .github/workflows/test.yml
Comment (michaelpj) [reply .github/workflows/test.yml:150]: Yeah, fair.

Comment (jneira) [reply .github/workflows/test.yml:150]: An alternative is mark them as ignored or broken for that hls version using tasty modifiers, but we are not using them that way for now

File: .github/workflows/hackage.yml
Comment (jneira) [reply .github/workflows/hackage.yml:114]: Hmm, not sure if i am understanding, those tweaks are already in cabal-ghc901.project but we dont want to use the entire file, only the things needed by the solver which we know will not make the hackage build fail.
For example we are not building brittany for ghc-9, neither in the project nor in hackage so it is safe to add the `allow-newer`'s of brittany to make happy the solver cause the hackage build for ghc-9 of brittany is disabled by flags. 

Another alternative could be comment out the not buildable packages (like we do temporary, see for example the actual `cabal-ghc921.project`) but it would make more complex the actual `echo "packages: . ../../* ../../plugins/*" > cabal.project`

Comment (michaelpj) [reply .github/workflows/hackage.yml:114]: Ah I see, I think I just didn't understand the point either. Let me think about it a bit, and anyway it's irrelevant to this PR :)

**Comment by wz1000** [comment]
We don't need a whole new `cabal.project` for GHC 9.0.2. The one for 9.0.1 will suffice with a few small tweaks.

**Comment by pepeiborra** [comment]
Is there any way to avoid introducing a `cabal-ghc902.project` file? I see two reasonable options:
- Keep it, and delete `cabal-ghc901.project`
- Rename `cabal-ghc901.project` to `cabal-ghc9.project` and make it work for both

**Comment by thomasjm** [comment]
Thanks for putting this together! FWIW I tried to build it with `stack build --stack-yaml stack-9.0.2.yaml` and it doesn't build. 

I had to add `unix` to `extra-deps`, and then it hit errors building `ghc-parser-ex`, `hie-bios`, `ormolu`, and `fourmolu`. Also a GHC panic building `primitive-unlifted`.

**Comment by jneira** [comment]
> Thanks for putting this together! FWIW I tried to build it with `stack build --stack-yaml stack-9.0.2.yaml` and it doesn't build.
> 
> I had to add `unix` to `extra-deps`, and then it hit errors building `ghc-parser-ex`, `hie-bios`, `ormolu`, and `fourmolu`. Also a GHC panic building `primitive-unlifted`. [Logs](https://pastebin.com/nzu9y0nh)

thank for trying, circleci is asking unix too, will try to fix it asap

**Comment by jneira** [comment]
> Is there any way to avoid introducing a `cabal-ghc902.project` file? I see two reasonable options:
> 
>     * Keep it, and delete `cabal-ghc901.project`
> 
>     * Rename `cabal-ghc901.project` to `cabal-ghc9.project` and make it work for both

i am for renaming to `cabal-ghc9.project`, there is a little chance of make the 9.0.1 build fail or fail in hackage due to 9.0.2 specific tweaks but i suppose is small enough to worht the deduplication

**Comment by jneira** [comment]
after adding unix ci reached the error commented by @thomasjm for primitive-unlifted:

```
primitive-unlifted                > [ 7 of 18] Compiling Data.Primitive.Unlifted.MVar.Primops
primitive-unlifted                > ghc: panic! (the 'impossible' happened)
primitive-unlifted                >   (GHC version 9.0.2:
primitive-unlifted                > 	findSlot
primitive-unlifted                >   Can't find slot
primitive-unlifted                >   [PtrUnliftedSlot]
primitive-unlifted                >   [PtrLiftedSlot]
primitive-unlifted                >   Call stack:
primitive-unlifted                >       CallStack (from HasCallStack):
primitive-unlifted                >         callStackDoc, called at compiler/GHC/Utils/Outputable.hs:1230:37 in ghc:GHC.Utils.Outputable
primitive-unlifted                >         pprPanic, called at compiler/GHC/Types/RepType.hs:217:9 in ghc:GHC.Types.RepType
primitive-unlifted                > 
primitive-unlifted                > Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug
```

**Comment by jneira** [comment]
> after adding unix ci reached the error commented by @thomasjm for primitive-unlifted:
> 

issue in the ghc issue tracker: https://gitlab.haskell.org/ghc/ghc/-/issues/20908 :worried:

**Comment by wz1000** [comment]
The solution is to use primitive-unlifted < 1.0

`primitive-unflifted-1.0` does some shady things which aren't semantically valid.

**Comment by jneira** [comment]
> The solution is to use primitive-unlifted < 1.0
> 

thanks for the suggestion, it is not a direct dependency of hls so i guess it will be hard to make the hackage build works although we can force that version using cabal.prokect/stack.yaml

**Comment by wz1000** [comment]
See https://github.com/metrix-ai/primitive-extras/issues/6

**Comment by jneira** [comment]
Ok we hit the hie-bios error with cabal:

```
 src/HIE/Bios/Ghc/Gap.hs:260:7: error:
Error:     • Couldn't match expected type ‘IncludeSpecs’
                  with actual type ‘[String] -> IncludeSpecs’
    • Probable cause: ‘IncludeSpecs’ is applied to too few arguments
      In the ‘includePaths’ field of a record
      In the expression:
        df
          {includePaths = IncludeSpecs
                            (map f $ G.includePathsQuote (includePaths df))
                            (map f $ G.includePathsGlobal (includePaths df))}
      In an equation for ‘mapOverIncludePaths’:
          mapOverIncludePaths f df
            = df
                {includePaths = IncludeSpecs
                                  (map f $ G.includePathsQuote (includePaths df))
                                  (map f $ G.includePathsGlobal (includePaths df))}
    |
260 |       G.IncludeSpecs
    |       ^^^^^^^^^^^^^^...
```
so we have to upgrade hie-bios

In stack we got another error, already pointed out by @thomasjm :

```
ghc-lib-parser-ex                 > /tmp/stack-bcedea5ffccdf320/ghc-lib-parser-ex-9.0.0.6/src/Language/Haskell/GhclibParserEx/GHC/Settings/Config.hs:18:1: error:
ghc-lib-parser-ex                 >     Could not load module ‘GHC.Platform’
ghc-lib-parser-ex                 >     It is a member of the hidden package ‘ghc-boot-9.0.2’.
ghc-lib-parser-ex                 >     Perhaps you need to add ‘ghc-boot’ to the build-depends in your .cabal file.
ghc-lib-parser-ex                 >     Use -v (or `:set -v` in ghci) to see a list of the files searched for.
ghc-lib-parser-ex                 >    |
ghc-lib-parser-ex                 > 18 | import GHC.Platform
ghc-lib-parser-ex                 >    | ^^^^^^^^^^^^^^^^^^^
```

**Comment by wz1000** [comment]
Can we also drop stack-9.0.1.yaml? GHC 9.0.1 is not a version anyone should be using.

**Comment by jneira** [comment]
> Can we also drop stack-9.0.1.yaml? GHC 9.0.1 is not a version anyone should be using.

well, we are planning remove all (but one or two) together and i would prefer to keep them consistent until that, having one stack.yaml for each supported ghc (we have also have stack-8.10.6.yaml). 
As noted previously it is needed to make work the install script and the 9.0.2 is still recent so users have to upgrade their instalaltions

**Comment by jneira** [comment]
ghc-lib-parser-ex fails for stack with:

```
ghc-lib-parser-ex                 > /tmp/stack-83379216794bc4a2/ghc-lib-parser-ex-9.0.0.6/src/Language/Haskell/GhclibParserEx/GHC/Settings/Config.hs:18:1: error:
ghc-lib-parser-ex                 >     Could not load module ‘GHC.Platform’
ghc-lib-parser-ex                 >     It is a member of the hidden package ‘ghc-boot-9.0.2’.
ghc-lib-parser-ex                 >     Perhaps you need to add ‘ghc-boot’ to the build-depends in your .cabal file.
ghc-lib-parser-ex                 >     Use -v (or `:set -v` in ghci) to see a list of the files searched for.
ghc-lib-parser-ex                 >    |
ghc-lib-parser-ex                 > 18 | import GHC.Platform
ghc-lib-parser-ex                 >    | ^^^^^^^^^^^^^^^^^^^
....
--  While building package ghc-lib-parser-ex-9.0.0.6 (scroll up to its section to see the error) using:
      /root/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_3.4.1.0_ghc-9.0.2 --builddir=.stack-work/dist/x86_64-linux/Cabal-3.4.1.0 build --ghc-options " -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
```

but cabal is building succesfully ghc-lib-parser-ex-9.0.0.5

**Comment by michaelpj** [comment]
> but cabal is building succesfully ghc-lib-parser-ex-9.0.0.5

That seems surprising since it's deprecated on hackage and the base bounds are set to be unsatisfiable :thinking:

**Comment by jneira** [comment]
> > but cabal is building succesfully ghc-lib-parser-ex-9.0.0.5
> 
> That seems surprising since it's deprecated on hackage and the base bounds are set to be unsatisfiable thinking

you may be right, i dont find the build where i thought i've seen it

**Comment by michaelpj** [comment]
I think you need to add `allow-newer: ghc-typelits-natnormalise:ghc-bignum` to `ghcide/test/data/plugin-knownnat/cabal.project`.

**Comment by jneira** [comment]
> I think you need to add `allow-newer: ghc-typelits-natnormalise:ghc-bignum` to `ghcide/test/data/plugin-knownnat/cabal.project`.

yeah, thanks for the suggestion, i wonder if we could change it to not have similar problems in the future

File: .github/workflows/build.yml
Comment (michaelpj) [inline .github/workflows/build.yml:58]: Wait, shouldn't this just be using the setup-build workflow from above?

File: cabal-ghc90.project
Comment (michaelpj) [inline cabal-ghc90.project:30]: I'm assuming this didn't matter? We won't have another GHC in scope or anything? (I don't think we will, but just checking)

File: .github/workflows/build.yml
Comment (jneira) [reply .github/workflows/build.yml:58]: yeah, it should but we were not using caching here, not sure why. I had that as TODO in my mind :-P

File: cabal-ghc90.project
Comment (jneira) [reply cabal-ghc90.project:30]: the line only saved you `-w ghc-9.0.1`. Now the file should work for both ghc 9.0.1 and 9.0.2 and that line would make 9.0.2 in PATH (like we do in ci) be ignored

File: .github/workflows/build.yml
Comment (jneira) [reply .github/workflows/build.yml:58]: my idea is check if caching should be effectively avoided and separate the actual action in two, extracting the cache part, in the other case

Comment (Anton-Latukha) [reply .github/workflows/build.yml:58]: Yes. Neither of us initially planne to refactor `build`. `build` seems like particularly rarely used & important workflow & I left it to the more experienced people. Consolidating & solidifying the work between `caching`, `test`, `bench` already was enough task & scope. We always can share the work done into `build`, but at the start that approach was ineffective (consolidating between 4 targets at once may be exponentially harder task).

File: .github/workflows/hackage.yml
Comment (Anton-Latukha) [reply .github/workflows/hackage.yml:107]: Just a *nota bene*: the https://github.com/haskell/haskell-language-server/pull/2567/files#r779448472 is indeed a complex reason & seems like a case for a code comment/doc "why it was needed here".

File: .github/workflows/hackage.yml
Comment (Anton-Latukha) [inline .github/workflows/hackage.yml:115]: Seems like the same reasoning from https://github.com/haskell/haskell-language-server/pull/2567/files#r779448472 to be documented

Seems like the sum of: https://github.com/haskell/haskell-language-server/pull/2567/files#r779448472 & https://github.com/haskell/haskell-language-server/pull/2567/files#r779120031 is enough to form a rough understanding.

But also `hackage` & `build` are a "deep cut" stuff for core maintainers, but even in that case doc may be beneficial to keep in the file.

File: .github/workflows/test.yml
Comment (Anton-Latukha) [reply .github/workflows/test.yml]: There is nothing ugly in the ugly shell script in the `.workflow` directory.

I would post an argument that supplying a function through `env` or through `.bashrc` would be dirtier.

Also, the GHC is in a build environment, so it can be `ghci` script also.

It would be:
```shell
#!/usr/bin/env sh

BUILD_TARGET_NAME=$1

TEST_OPTIONS=$2

cabal test "$BUILD_TARGET_NAME" --test-options="$TEST_OPTIONS" || cabal test "$BUILD_TARGET_NAME" --test-options="$TEST_OPTIONS" || LSP_TEST_LOG_COLOR=0 LSP_TEST_LOG_MESSAGES=true LSP_TEST_LOG_STDERR=true cabal test "$BUILD_TARGET_NAME" --test-options="$TEST_OPTIONS"
```

```shell
[pyro@rogue haskell-language-server]$ bash -x ./draft.sh hls-brittany-plugin TOPT
+ BUILD_TARGET_NAME=hls-brittany-plugin
+ TEST_OPTIONS=TOPT
+ cabal test hls-brittany-plugin --test-options=TOPT
Resolving dependencies...
```

But I think this PR is already to big. Just an idea to refactor after merge.

Comment (Anton-Latukha) [reply .github/workflows/test.yml]: Reported in https://github.com/haskell/haskell-language-server/issues/2586

File: .github/workflows/hackage.yml
Comment (Anton-Latukha) [reply .github/workflows/hackage.yml:107]: Reported in https://github.com/haskell/haskell-language-server/issues/2586

File: .github/workflows/build.yml
Comment (Anton-Latukha) [reply .github/workflows/build.yml:58]: Reported in https://github.com/haskell/haskell-language-server/issues/2586

**Comment by Anton-Latukha** [comment]
This `+0.0.1` update is a doozy. It is already big enough, the comments we made that are not related to the agenda - gathered into #2586.

File: .github/workflows/caching.yml
Comment (Anton-Latukha) [inline .github/workflows/caching.yml:48]: *nota bene*, commit message explains this, this indeed needed to be removed currently, to simplify things to workaround hard `cabal` tooling/package description design difficulties.

File: .github/workflows/caching.yml
Comment (Anton-Latukha) [inline .github/workflows/caching.yml:50]: *nota bene* commit message explains this, this indeed needed to be removed currently, to simplify things to workaround hard `cabal` tooling/package description design difficulties.

Comment (jneira) [reply .github/workflows/caching.yml:50]: agree, will add the commit msg in the file

**Comment by jneira** [comment]
As expected the caching workflow was succesful: https://github.com/jneira/haskell-language-server/runs/4804786986?check_suite_focus=true

**Comment by jneira** [comment]
@michaelpj @Anton-Latukha I ve changed the caching workflow to build bench deps adding some comments (and in other workflow filed) hope it makes thing a little bit clearer

**Review by Anton-Latukha — APPROVED** [review]
I was Ok with previous (pre `9.2.1` changes merge), & I am Ok with the recently mentioned `caching` changes.

**Comment by jneira** [comment]
The new caching workflow has been succesful here: https://github.com/jneira/haskell-language-server/actions/runs/1704688585

File: .github/workflows/build.yml
Comment (jneira) [inline .github/workflows/build.yml:25]: Unnecessary change
```suggestion
          , "8.10.7"
```

**Review by michaelpj — APPROVED** [review]

**Pull request merged**
